libass/ass.c

```diff
@@ -1455,18 +1455,22 @@ char *ass_load_file(ASS_Library *library, const char *fname, FileNameSource hint
         return NULL;
     }
 
-    const size_t CHUNK_SIZE = 16 * 1024;
-    size_t bufmax = 128 * CHUNK_SIZE;
-    char *buf = malloc(bufmax + 1); // one extra for terminating zero
+    char *buf = NULL;
+    size_t bufmax = 2 * 1024 * 1024;
+    ass_get_file_size(fp, &bufmax);
+    if (bufmax > SIZE_MAX - 1 || bufmax == 0)
```

Comment:
Small `bufmax` can lead to suboptimal realloc scaling (i.e. 1, 2, 4,...), so I suggest to do `bufmax += CHUNK_SIZE` before allocation.

Reply:
I'm suggesting to read not only one additional byte after EOF, but `CHUNK_SIZE` additional bytes. It's to guarantee that any `fread` tries to read at least `CHUNK_SIZE` bytes for some sensible `CHUNK_SIZE`. Without that, there can be some miniscule reads immediately before reallocs in case of incremental processing. So yes, I'm suggesting to keep `CHUNK_SIZE` and old realloc condition (but new `fread` size is better, given that `maxread >= CHUNK_SIZE`).